### ISSUE TITLE
Sometimes caret does not move to new paragraph on enter

### ISSUE THREAD

**Issue by iseulde** [WordPress/gutenberg#10484]
**Describe the bug**

This is an issue after #10439, reported by @aduth: https://github.com/WordPress/gutenberg/pull/10439#pullrequestreview-163391445

**To Reproduce**

> 1. New post
> 2. Click writing prompt
> 3. Cmd+B to bold
> 4. Type something
> 5. Press Enter
> 6. Press Backspace
> 7. Press Enter
> 
> Expected: Typing in second paragraph.
> Actual: Second paragraph created, but caret remains in first.
> 
> Works as expected in 3.9.0.
> Should have E2E case (could extend "should merge into inline boundary position")

**Comment by iseulde** [comment]
The problem as far as I know comes from not being able to pass new instances of strings, so the string value is the same when compared in component did update, where the caret position is retired on merge. Visually it still seems to be restored correctly, but after the inline boundary position if there is formatting. If enter is then pressed, the caret does not moved. I have not figured out yet why that happens.

In any case, this bug only happens _sometimes_ when pressing enter from the _outside of a boundary position_.
